Signal/src/ViewControllers/MediaPageViewController.swift

```diff
@@ -0,0 +1,706 @@
+            return nil
+        }
+
+        guard let previousItem = buildGalleryItem(mediaMessage: previousMediaMessage, thread: thread) else {
+            return nil
+        }
+
+        cachedItems.insert(previousItem, at: currentIndex)
+        return previousItem.viewController
+    }
+
+    public func pageViewController(_ pageViewController: UIPageViewController, viewControllerAfter viewController: UIViewController) -> UIViewController? {
+        Logger.debug("\(logTag) in \(#function)")
+
+        guard let currentIndex = cachedItems.index(where: { $0.viewController == viewController }) else {
+            owsFail("\(self.logTag) unknown view controller. \(viewController)")
+            return nil
+        }
+        let currentItem = cachedItems[currentIndex]
+
+        let newIndex = currentIndex + 1
+        if let cachedItem = cachedItems[safe: newIndex] {
+            return cachedItem.viewController
+        }
+
+        guard let nextMediaMessage = nextMediaMessage(currentItem.message) else {
+            return nil
+        }
+
+        guard let nextItem = buildGalleryItem(mediaMessage: nextMediaMessage, thread: thread) else {
+            return nil
+        }
+
+        cachedItems.insert(nextItem, at: newIndex)
+        return nextItem.viewController
+    }
+
+    private func buildGalleryItem(mediaMessage: TSMessage, thread: TSThread) -> MediaGalleryItem? {
+        var fetchedAttachment: TSAttachment? = nil
+        var fetchedItem: ConversationViewItem? = nil
+        self.uiDatabaseConnection.read { transaction in
+            fetchedAttachment = mediaMessage.attachment(with: transaction)
+            fetchedItem = ConversationViewItem(interaction: mediaMessage, isGroupThread: thread.isGroupThread(), transaction: transaction)
+        }
+
+        guard let attachmentStream = fetchedAttachment as? TSAttachmentStream else {
+            owsFail("attachment stream unexpectedly nil")
+            return nil
+        }
+
+        guard let viewItem = fetchedItem else {
+            owsFail("viewItem stream unexpectedly nil")
+            return nil
+        }
+
+        let viewController = MediaDetailViewController(attachmentStream: attachmentStream, viewItem: viewItem)
+        viewController.delegate = self
+        return MediaGalleryItem(message: mediaMessage,
+                                attachmentStream: attachmentStream,
+                                viewController: viewController)
+    }
+
+    @nonobjc
+    public func presentationCount(for: UIPageViewController) -> Int {
+        Logger.debug("\(logTag) in \(#function)")
+
+        var count: UInt = 0
+        self.uiDatabaseConnection.read { (transaction: YapDatabaseReadTransaction) in
+            count = self.mediaGalleryFinder.mediaCount(thread: self.thread, transaction: transaction)
+        }
+        return Int(count)
+    }
+
+    @nonobjc
+    public func presentationIndex(for: UIPageViewController) -> Int {
```

Comment:
Could we do the `presentationIndex(for page:` instead to avoid the "for" funny business?

Reply:
Yes, much better, thanks!